File: frontend-miniapp/miniprogram/features/message/data/message-api.ts
import { apiRequest } from '../../../core/http/api-client'
import type {
  ChatMessage,
  ChatSession,
  CreateSessionRequest,
  NotificationItem,
  SendMessageRequest,
} from '../domain/message-models'

function withQuery(path: string, params: Record<string, string | number | undefined>): string {
  const parts: string[] = []
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === '') continue
    parts.push(`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`)
  }
  return parts.length ? `${path}?${parts.join('&')}` : path
}

export const chatSessionApi = {
  create(payload: CreateSessionRequest): Promise<ChatSession> {
    return apiRequest({ path: '/api/chat/sessions', method: 'POST', data: payload })
  },

  listMy(): Promise<ChatSession[]> {
    return apiRequest({ path: '/api/chat/sessions' })
  },

  listCs(): Promise<ChatSession[]> {
    return apiRequest({ path: '/api/chat/sessions/cs' })
  },

  assign(sessionId: string): Promise<ChatSession> {
    return apiRequest({ path: `/api/chat/sessions/${sessionId}/assign`, method: 'PUT' })
  },
}

export const chatMessageApi = {
  send(sessionId: string, payload: SendMessageRequest): Promise<ChatMessage> {
    return apiRequest({ path: `/api/chat/messages/${sessionId}`, method: 'POST', data: payload })
  },

  list(sessionId: string, page = 1, pageSize = 50): Promise<ChatMessage[]> {
    return apiRequest({ path: withQuery(`/api/chat/messages/${sessionId}`, { page, pageSize }) })
  },
}

export const notificationApi = {
  list(category?: number, page = 1, pageSize = 20): Promise<NotificationItem[]> {
    return apiRequest({
      path: withQuery('/api/message/notifications', { category, page, pageSize }),
    })
  },

  markRead(notificationId: string): Promise<void> {
    return apiRequest({ path: `/api/message/notifications/${notificationId}/read`, method: 'PUT' })
  },

  markAllRead(): Promise<void> {
    return apiRequest({ path: '/api/message/notifications/read-all', method: 'PUT' })
  },
}

File: frontend-app/lib/features/message/data/message_api.dart
import 'package:shopping_app/core/network/api_client.dart';
import 'package:shopping_app/core/network/api_response.dart';
import 'package:shopping_app/features/message/domain/message_models.dart';

class MessageApi {
  MessageApi(this._client);

  final ApiClient _client;

  Future<ChatSession> createSession({
    String? shopId,
    String? subject,
    String? firstMessage,
  }) async {
    final response = await _client.post(
      '/api/chat/sessions',
      data: {
        if (shopId != null && shopId.isNotEmpty) 'shopId': shopId,
        if (subject != null && subject.isNotEmpty) 'subject': subject,
        if (firstMessage != null && firstMessage.isNotEmpty)
          'firstMessage': firstMessage,
      },
    );
    return unwrapApiResponse(response.data, ChatSession.fromJson);
  }

  Future<List<ChatSession>> listMySessions() async {
    final response = await _client.get('/api/chat/sessions');
    return unwrapApiResponse(
      response.data,
      (value) => readObjectList(value, ChatSession.fromJson, message: '会话列表格式错误'),
    );
  }

  Future<List<ChatSession>> listCsSessions() async {
    final response = await _client.get('/api/chat/sessions/cs');
    return unwrapApiResponse(
      response.data,
      (value) => readObjectList(value, ChatSession.fromJson, message: '客服会话列表格式错误'),
    );
  }

  Future<ChatSession> assignSession(String sessionId) async {
    final response = await _client.put('/api/chat/sessions/$sessionId/assign');
    return unwrapApiResponse(response.data, ChatSession.fromJson);
  }

  Future<void> closeSession(String sessionId) async {
    final response = await _client.put('/api/chat/sessions/$sessionId/close');
    unwrapApiResponse<void>(response.data, (_) {});
  }

  Future<List<ChatMessage>> listMessages(
    String sessionId, {
    int page = 1,
    int pageSize = 50,
  }) async {
    final response = await _client.get(
      '/api/chat/messages/$sessionId',
      queryParameters: {'page': page, 'pageSize': pageSize},
    );
    return unwrapApiResponse(
      response.data,
      (value) => readObjectList(value, ChatMessage.fromJson, message: '消息列表格式错误'),
    );
  }

  Future<ChatMessage> sendMessage(String sessionId, String content) async {
    final response = await _client.post(
      '/api/chat/messages/$sessionId',
      data: {'content': content, 'msgType': 1},
    );
    return unwrapApiResponse(response.data, ChatMessage.fromJson);
  }

  Future<List<NotificationItem>> listNotifications({
    int page = 1,
    int pageSize = 20,
  }) async {
    final response = await _client.get(
      '/api/message/notifications',
      queryParameters: {'page': page, 'pageSize': pageSize},
    );
    return unwrapApiResponse(
      response.data,
      (value) =>
          readObjectList(value, NotificationItem.fromJson, message: '通知列表格式错误'),
    );
  }

  Future<void> markNotificationRead(String notificationId) async {
    final response = await _client.put(
      '/api/message/notifications/$notificationId/read',
    );
    unwrapApiResponse<void>(response.data, (_) {});
  }

  Future<void> markAllNotificationsRead() async {
    final response = await _client.put('/api/message/notifications/read-all');
    unwrapApiResponse<void>(response.data, (_) {});
  }
}
